### ISSUE TITLE
Implement toString for InputTypes

### ISSUE THREAD

**Issue by robertsonmcclure** [apollographql/apollo-kotlin#3143]
Java classes that are generated for the response of a GraphQL query/mutation have really nice toString methods. However, the classes generated for InputTypes do not have a toString (i.e. https://github.com/apollographql/apollo-android/blob/f72c3afd17655591aca90a6a118dbb7be9c50920/apollo-compiler/src/test/graphql/com/example/input_object_type/type/ColorInput.java). Having a toString for Input classes would make it easier to print the arguments that I'm calling my GraphQL Server with, especially if they are nested. 

Since I am not sure exactly how to do this, I would appreciate some pointers on how to implement this.

**Comment by martinbonnin** [comment]
Thanks for reaching out! Is Kotlin codegen an option? IIRC, Kotlin input types are data classes so you will get `toString()` for free by setting `generateKotlinModels.set(true)`

**Comment by robertsonmcclure** [comment]
Potentially, but what if using Kotlin codegen wasn't an option?

**Comment by martinbonnin** [comment]
You'd have to write the javapoet code to add `toString()` like there is `hashCode()` and `equals()` here: https://github.com/apollographql/apollo-android/blob/main/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/InputTypeSpecBuilder.kt#L27

**Comment by robertsonmcclure** [comment]
I looked through the other TypeSpec files in the folder that has InputTypeSpecBuilder, and I found a place where `.withToStringImplementation()` is used: https://github.com/apollographql/apollo-android/blob/f72c3afd17655591aca90a6a118dbb7be9c50920/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/SchemaTypeSpecBuilder.kt#L118

I couldn't figure out if it was referencing anything, as searching the codebase didn't turn up anything that looked like an implementation of toString for SchemaTypeSpecBuilder. 

Would this change consist of a one line line addition of `.withToStringImplementation()`, or am I misunderstanding something?

**Comment by martinbonnin** [comment]
> Would this change consist of a one line line addition of .withToStringImplementation(), or am I misunderstanding something?

I'm not overly familiar with the java codegen but that might be it! Want to send a PR?

**Comment by robertsonmcclure** [comment]
Are there step by step instructions on how to build this project locally & run the tests? I tried but ran into issues with gradle & IntelliJ. 

I know the change that I want to make, but I would like to make sure the build passes, existing tests pass, and add a test or two for the new functionality.

**Comment by robertsonmcclure** [comment]
Also, I upgrade to generating Kotlin classes instead of Java classes, and ran into a similar but different issue. The `data class` generates a good toString for all fields that are mandatory, however the Input class is not a data class and does not have a good toString. 

```
CreateCaseInput(caseId=com.apollographql.apollo.api.Input@4d5, tenantId=testTenantId, entityType=defaultEntityType, entityId=aabbd872-6057-
422c-9849-b8930df59820, status=defaultStatus, owner=com.apollographql.apollo.api.Input@4cf, marketplaceId=com.apollographql.apollo.api.Input@4d5, sla=com.apollographql.apollo.api.Input@4cf, createdByAgent=com.apollographql.apollo.api.Input@1a96700b, deduplicationId=com.ap
ollographql.apollo.api.Input@4d5, propertyValues=com.apollographql.apollo.api.Input@b11c90fc, investigations=com.apollographql.apollo.api.Input@4d5, createdAt=com.apollographql.apollo.api.Input@4cf, createdBySystem=com.apollographql.apollo.api.Input@4d5, updatedAt=com.apo
llographql.apollo.api.Input@4cf, updatedByAgent=com.apollographql.apollo.api.Input@4d5, updatedBySystem=com.apollographql.apollo.api.Input@4d5)
```

**Comment by martinbonnin** [comment]
> Are there step by step instructions on how to build this project locally & run the tests? 

We do have a [Contributing.md](https://github.com/apollographql/apollo-android/blob/main/Contributing.md) but it's unfortunately not very detailed. The main reason is exactly what you bumped into that IntelliJ support has a few issues so autocomplete doesn't always work, etc. I'm using the latest IntelliJ but didn't upgrade the Kotlin plugin (so I'm using IntelliJ Kotlin plugin 1.4.32). And the setup changes quite often so it's hard to track what works best. I'd say if you have the option, prefer IntelliJ to Android Studio though.

To run the tests, the [CI script](https://github.com/apollographql/apollo-android/blob/main/scripts/ci.sh) is the source of truth.

The tests can be run with `./gradlew -p composite build`.

Since you're touching codegen, the important tests are compiler tests. Run the below to update them:

```
./gradlew :apollo-compiler:test -DupdateTestFixtures=true
```

> the Input class is not a data class and does not have a good toString.

I see. I wonder if this is just an oversight or if there are reasons it was not included. Since both `equals` and `hashCode` are there, it seems like `toString` would be a good fit but maybe I'm missing something here.

**Comment by martinbonnin** [comment]
Implemented as of `3.8.4`
